pkg/lib/cockpit-components-install-dialog.jsx

```diff
@@ -0,0 +1,201 @@
+    }
+
+    function check_missing() {
+        PK.check_missing_packages([ pkg ],
+                                  p => {
+                                      cancel = p.cancel;
+                                      var pm = null;
+                                      if (p.waiting)
+                                          pm = _("Waiting for other software management operations to finish");
+                                      else
+                                          pm = _("Checking installed software");
+                                      if (pm != progress_message) {
+                                          progress_message = pm;
+                                          update();
+                                      }
+                                  })
+                .then(d => {
+                    if (d.unavailable_names.length > 0)
+                        error_message = cockpit.format(_("$0 is not available from any repository."),
+                                                       d.unavailable_names[0]);
+                    else
+                        data = d;
+                    progress_message = null;
+                    cancel = null;
+                    update();
+                })
+                .catch(e => {
+                    progress_message = null;
+                    cancel = null;
+                    error_message = e.toString();
+                    update();
+                });
+    }
+
+    function install_missing() {
+        // We need to return a Cockpit flavoured promise since we want
+        // to use progress notifications.
+        var dfd = cockpit.defer();
+
+        PK.install_missing_packages(data,
+                                    p => {
+                                        var text = null;
+                                        if (p.waiting) {
+                                            text = _("Waiting for other software management operations to finish");
+                                        } else if (p.package) {
+                                            var fmt;
+                                            if (p.info == PK.Enum.INFO_DOWNLOADING)
+                                                fmt = _("Downloading $0");
+                                            else if (p.info == PK.Enum.INFO_REMOVING)
+                                                fmt = _("Removing $0");
+                                            else
+                                                fmt = _("Installing $0");
+                                            text = format_to_array(fmt, <strong>{p.package}</strong>);
+                                        }
+                                        dfd.notify(text, p.cancel);
+                                    })
+                .then(() => {
+                    dfd.resolve();
+                })
+                .catch(error => {
+                    dfd.reject(error);
+                });
```

Comment:
Note: should suffice to write just `.then(dfd.resolve)` and `.catch(dfd.reject)`; but it's fine like that. Does the caller firmly expect a cockpit promise? A standard Promise doesn't work?

Reply:
I am not convinced that the two forms are equivalent, because of `this`.  With a call like `dfd.resolve()`, the resolve method will have `this == dfd`.  With just `.then(dfd.resolve)`, `this` is likely undefined.  This might not matter depending on whether the resolve method actually uses `this` or not.  I think it doesn't, but that's just luck.  Am I wrong?

Reply:
Ah, could be. We use that pattern a lot (`git grep fail.*reject`) and it seems to work, but indeed it could be that this is by luck.